### PR TITLE
inList empty list argument support  

### PR DESCRIPTION
**Pull request by demon101**
empty list support for InList and NotInList

### CODE REVIEW

**Comment by demon101** [comment]
https://github.com/grails/grails-data-mapping/issues/1212

File: grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/finders/MethodExpression.java [hunk 1]
```diff
@@ -258,7 +264,12 @@ public InList(String propertyName) {
 
         @Override
         public Query.Criterion createCriterion() {
-            return Restrictions.in(propertyName, (Collection<?>) arguments[0]);
+            Collection<?> argument = (Collection<?>) arguments[0];
+            if (argument == null || argument.isEmpty()) {
+                return Restrictions.and(Restrictions.isNull(propertyName), Restrictions.isNotNull(propertyName));
```
Comment (jeffbrown) [inline grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/finders/MethodExpression.java]: I don't understand why `Restrictions.and(Restrictions.isNull(propertyName), Restrictions.isNotNull(propertyName))` would make sense.  Will that always effectively be `false`?  If so, surely there is a better way than adding to the criteria these 2 conditions.

Comment (demon101) [reply grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/finders/MethodExpression.java]: Yes, always false. But I don't know how to do it better. 
https://github.com/grails/grails-data-mapping/issues/1212#issuecomment-455879135

Comment (demon101) [reply grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/finders/MethodExpression.java]: Dear @jeffbrown , I would be very grateful if you could tell me how to make it easier. I have not found easy solution

**Comment by jameskleeh** [comment]
Adding an always false criteria to the query is really a workaround for not executing the query at all. If the query will always return an empty set, then why make the trip to the DB at all?

**Comment by demon101** [comment]
I can add method to MethodExpression like "isEmptyResult" and call it before query building, but I have to send result of the cunction upper, 3 levels in the stack 
or I can use null result value for MethodExpression.createCriterion , also not nice
another way -- extend Query class, but as I understend, it can produce problems with query cashing
best way (IMHO) use some universal logical false expression

**Comment by graemerocher** [comment]
Yeah it just seems wasteful to execute the query at all in this case, hence why we are unsure about merging the current state of the PR

**Comment by demon101** [comment]
@graemerocher I wasn't sure about the solution with extra query. And I asked about https://github.com/grails/grails-data-mapping/issues/1212#issuecomment-455879135 .   But for me, extra call better then exception. 

I can't see simple good solutions.  Query or dynamic finder class logic have to be changed. 
If you can show me general direction of solution, I can implement it.

**Comment by jeffbrown** [comment]
> Adding an always false criteria to the query is really a workaround for not executing the query at all. If the query will always return an empty set, then why make the trip to the DB at all?

I don't think the proposed solution is great but to address that question, the expression could be one side of an `OR` condition so I think shortcutting the round trip to the database any time this scenario occurs would be problematic.

**Comment by demon101** [comment]
> I don't think the proposed solution is great but to address that question, the expression could be one side of an `OR` condition so I think shortcutting the round trip to the database any time this scenario occurs would be problematic.

'OR' condition presented only in NotInList, InList fix implemented with 'AND'  condition. I can remove  fix for NotInList

**Comment by demon101** [comment]
> the expression could be one side of an `OR` condition so I think shortcutting the round trip to the database any time this scenario occurs would be problematic.

Yes, skipping of empty parts of query looks like query optimisation on GORM side. For me, extra query looks better then query with exception as result (as it is now). 

But expression `field = null and field !=null` is a real hack, since true=false not possible to build with gorm `Restrictions` (As I can see)

**Comment by zyro23** [comment]
is the proposed fix only targeting gorm for sql/hibernate datastores? i think other non-sql databases gorm supports (e.g. mongo) may be just fine with an empty collection argument?

n.b.: spring-data-jpa seems to work around this with `IN (NULL)`. that also always evaluates to false because `NULL` in sql is never equal.

* https://jira.spring.io/browse/DATAJPA-606
* https://github.com/spring-projects/spring-data-jpa/blob/2.1.6.RELEASE/src/main/java/org/springframework/data/jpa/provider/PersistenceProvider.java#L103
* https://hibernate.atlassian.net/browse/HHH-8091

**Comment by demon101** [comment]
> i think other non-sql databases gorm supports (e.g. mongo) may be just fine with an empty collection argument?

Yes, right!

I think, better to move hack to `in` implementation of sql/hibernate datastores. 
@jeffbrown I can make new PR if it satisfies the grails command

**Comment by demon101** [comment]
new PR created 
https://github.com/grails/grails-data-mapping/pull/1247

`IN (NULL)`  only for JPA

**Pull request closed without merging**